### PR TITLE
Address-Only Topics and Queues

### PR DESCRIPTION
**Pull request by slang25**
This change adds 3 new builders
- `QueueAddressSubscriptionBuilder`: subscribing to a queue from an address (either QueueURL or ARN)
- `TopicAddressPublicationBuilder`: publishing to a topic from an address (ARN)
- `QueueAddressPublicationBuilder`: publishing message directly to a queue (either QueueURL or ARN)

This adds an infrastructure only mode, striking a balance between reuse of existing JustSaying concepts and code, while keeping away from some of the existing complexities around managing infrastructure.

This PR also cleans up some of the legacy code, separating out the publishers from the sqs & sns infrastructure classes, and reducing some of the redundant class hierarchy.

TODO:
- Update docs
- Get feedback

Things that deliberately haven't been done yet, but could added later:
- Add a validation mode that can check for existence of topic/queue, and possibly a configurable subscription check.

Examples of subscribing to pre-existing queue:
```csharp
config.Subscriptions(x =>
{
    x.ForQueueArn<OrderPlacedEvent>("arn:aws:sqs:eu-west-1:111122223333:queue1");
    
    // try the simplified builder
    x.ForQueueArn<OrderPlacedEvent>("arn:aws:sqs:eu-west-1:111122223333:queue1", cfg => { });
    
    // From a queue url, the region can be inferred
    x.ForQueueUrl<OrderPlacedEvent>("https://sqs.eu-west-1.amazonaws.com/111122223333/queue1");
    
    // For localstack, you don't need to specify the region, but you can
    x.ForQueueUrl<OrderPlacedEvent>("http://localhost:4566/123456789012/queue1", "us-east-1");
    
    // For interop with the classic JustSaying Infrastructure
    var addressProvider = new AccountAddressProvider("111122223333", "eu-west-1");
    
    // Using `DefaultNamingConventions`, other implementations can be passed into the constructor of `AccountAddressProvider`
    x.ForQueueUri<OrderPlacedEvent>(addressProvider.GetQueueUriByConvention<OrderPlacedEvent>());
    // Explicit names
    x.ForQueueUri<OrderPlacedEvent>(addressProvider.GetQueueUri("queue1"));
}
```

### CODE REVIEW

File: samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs
Comment (martincostello) [inline samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:107]: Just spitballing without digging further, what about as well or instead for the signature:

```csharp
x.ForQueueArn<OrderPlacedEvent>("arn:aws:sqs:eu-west-1:111122223333:queue1")
```

Just thinking from an API usability/readability point-of-view.

File: samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs
Comment (martincostello) [inline samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:114]: Same here - also a variant for `Uri`?

File: src/JustSaying/AwsTools/MessageHandling/SnsMessagePublisher.cs
Comment (martincostello) [inline src/JustSaying/AwsTools/MessageHandling/SnsMessagePublisher.cs:23]: 😄

File: src/JustSaying/AwsTools/MessageHandling/SnsTopicByName.cs
Comment (martincostello) [inline src/JustSaying/AwsTools/MessageHandling/SnsTopicByName.cs:15]: 👍

File: src/JustSaying/AwsTools/MessageHandling/SnsTopicByName.cs
Comment (martincostello) [inline src/JustSaying/AwsTools/MessageHandling/SnsTopicByName.cs:44]: Nit: `snsWriteConfiguration` needs a null check/guard of some kind as this would NRE.

File: src/JustSaying/Fluent/AccountAddressProvider.cs
Comment (martincostello) [inline src/JustSaying/Fluent/AccountAddressProvider.cs:22]: Just an idea (as long as the public API surface area doesn't explode with possibilities), maybe also expose the ability to directly pass in a `RegionEndpoint`?

File: src/JustSaying/Fluent/AccountAddressProvider.cs
Comment (martincostello) [inline src/JustSaying/Fluent/AccountAddressProvider.cs:95]: Might be nicer to use `UriBuilder` to do this, then you'd only need to concat the host.

File: src/JustSaying/Fluent/QueueAddress.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddress.cs:64]: Nit: I think it would read better to invert the length check and throw first, rather than have it indented above like this.

File: src/JustSaying/Fluent/QueueAddress.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddress.cs:77]: Case insensitive?

File: src/JustSaying/Fluent/QueueAddress.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddress.cs:83]: `UriBuilder`

File: src/JustSaying/Fluent/QueueAddressQueue.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddressQueue.cs:17]: Could you use `Uri.Segments` to do this rather than split the string, or does that return too many bits?

File: src/JustSaying/Fluent/QueueAddressSubscriptionBuilder`1.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddressSubscriptionBuilder`1.cs:58]: _Maybe_ we'd want to use `typeof(T).Name`?

File: src/JustSaying/Fluent/QueueAddressSubscriptionBuilder`1.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddressSubscriptionBuilder`1.cs:65]: Same here - I forget if this makes it format the assembly qualified name or not.

File: src/JustSaying/Fluent/QueueAddressSubscriptionBuilder`1.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddressSubscriptionBuilder`1.cs:80]: +1

File: src/JustSaying/Fluent/TopicAddress.cs
Comment (martincostello) [inline src/JustSaying/Fluent/TopicAddress.cs:40]: Case insensitive?

File: src/JustSaying/Fluent/TopicAddressPublicationBuilder`1.cs
Comment (martincostello) [inline src/JustSaying/Fluent/TopicAddressPublicationBuilder`1.cs:31]: Same comments around `typeof(T)` in this file too.

File: src/JustSaying/Fluent/QueueAddressQueue.cs
Comment (slang25) [reply src/JustSaying/Fluent/QueueAddressQueue.cs:17]: I forgot all about `Uri.Segments`, will update

File: src/JustSaying/Fluent/QueueAddressSubscriptionBuilder`1.cs
Comment (slang25) [reply src/JustSaying/Fluent/QueueAddressSubscriptionBuilder`1.cs:65]: They will get ToStringed, so for our purposes it's the same as `Fullname` (i.e. `Namespace.TypeName`), but I think we aren't being constant across the codebase, there are places we using `ToString`, `Fullname` and `Name` for seemingly the same thing.

Comment (slang25) [reply src/JustSaying/Fluent/QueueAddressSubscriptionBuilder`1.cs:65]: https://docs.microsoft.com/en-us/dotnet/api/system.type.fullname?view=net-5.0#code-try-3
^ differences between how they behave

File: src/JustSaying/Fluent/QueueAddress.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddress.cs:43]: Should this now be called `FromUri` maybe?

File: src/JustSaying/Fluent/QueueAddress.cs
Comment (slang25) [reply src/JustSaying/Fluent/QueueAddress.cs:62]: This line would also be hit when the region couldn't be parsed by `RegionEndpoint.GetBySystemName`, but actually that could do with it's own message.

File: samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs
Comment (slang25) [reply samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:107]: Yeah, I'm with you on this, I think it's a little clearer rather than with overloads.
Would like to hear other views.

File: src/JustSaying/Fluent/AccountAddressProvider.cs
Comment (slang25) [reply src/JustSaying/Fluent/AccountAddressProvider.cs]: This made me chuckle:
https://github.com/aws/aws-sdk-net/blob/475822dec5e87954b7a47ac65995714ae1f1b115/sdk/src/Services/S3Control/Custom/Internal/S3ArnUtils/S3OutpostResource.cs#L148-L149

File: samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs
Comment (slang25) [inline samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:121]: @martincostello I've added top-level APIs (Option 2 here).
We could go either way with this, as it's mostly an aesthetic change.

Option 1 Pros:
Having a `QueueAddress.None` is handy, as we can use it as a signal to use the classic auto-generated infra for convenience, so going between the modes feels a little simpler rather than calling 2 different overloads (minor nit IMO)

Option 2 Pros:
Less method overload headache for consumers.
Shorter registration lines.

File: src/JustSaying/Fluent/QueueAddress.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddress.cs]:
```suggestion
                if (hostParts.Length < 2) throw new ArgumentException($"Could not infer region from {nameof(queueUri)}, please specify the region using the {nameof(regionName)} argument. If you are using localstack, the default region is us-east-1.");
```

File: src/JustSaying/Fluent/QueueAddress.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddress.cs:53]:
```suggestion
                if (!string.Equals(servicePart, "sqs", StringComparison.OrdinalIgnoreCase)) throw new ArgumentException("Must be an ARN for an SQS queue.", nameof(queueUri));
```

File: src/JustSaying/Fluent/QueueAddress.cs
Comment (martincostello) [inline src/JustSaying/Fluent/QueueAddress.cs:56]:
```suggestion
                var regionEndpoint = RegionEndpoint.GetBySystemName(regionHostPart) ?? throw new ArgumentException($"Could not parse region: {regionHostPart}", nameof(regionHostPart));
```

File: tests/JustSaying.UnitTests/Fluent/QueueAddressTests.cs
Comment (martincostello) [inline tests/JustSaying.UnitTests/Fluent/QueueAddressTests.cs:27]:
```suggestion
            Assert.Throws<ArgumentException>("whatevertheparameternameishere", () => QueueAddress.FromArn(null));
```

File: samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs
Comment (martincostello) [reply samples/src/JustSaying.Sample.Restaurant.KitchenConsole/Program.cs:121]: Out of the two on pure aesthetics, I'd go with 2.

File: tests/JustSaying.UnitTests/Fluent/QueueAddressTests.cs
Comment (slang25) [reply tests/JustSaying.UnitTests/Fluent/QueueAddressTests.cs:27]: I did not know that!

File: src/JustSaying/AwsTools/MessageHandling/SqsQueueByNameBase.cs
Comment (martincostello) [inline src/JustSaying/AwsTools/MessageHandling/SqsQueueByNameBase.cs:42]: Should we add `CancellationToken` parameters to these various methods?

Comment (slang25) [reply src/JustSaying/AwsTools/MessageHandling/SqsQueueByNameBase.cs:42]: We should, I'll hold off for a separate PR though, this ones getting a little heavy change wise

**Comment by slang25** [comment]
For reasons I don't understand, we often get cross test output from this line here:
https://github.com/slang25/JustSaying/blob/opt-infra2/tests/JustSaying.IntegrationTests/Logging/LogContextTests.cs#L45

Could do with a second pair of eyes on this tomorrow if anyone is free.

File: src/JustSaying/Fluent/AccountAddressProvider.cs
Comment (adammorr) [inline src/JustSaying/Fluent/AccountAddressProvider.cs:86]: I guess if we move "sqs" somewhere centrally then this could also be moved

**Review by adammorr — CHANGES_REQUESTED** [review]

File: src/JustSaying/Fluent/SubscriptionsBuilder.cs
Comment (adammorr) [reply src/JustSaying/Fluent/SubscriptionsBuilder.cs:207]: Urgh sorry I just saw your TODOs 🤦

**Comment by adammorr** [comment]
The approach looks good to me 👍

File: src/JustSaying/AwsTools/MessageHandling/SqsQueueByNameBase.cs
Comment (slang25) [reply src/JustSaying/AwsTools/MessageHandling/SqsQueueByNameBase.cs:20]: I've changed the other occurrence to be consistent with this. This code has been moved from `SqsQueueBase`, I'm trying to keep changes to a minimum while moving it about.

File: src/JustSaying/AwsTools/MessageHandling/SqsQueueByNameBase.cs
Comment (slang25) [reply src/JustSaying/AwsTools/MessageHandling/SqsQueueByNameBase.cs:218]: Probably, again this is moving existing code around, it could probably be improved but I'm hesitant to tackle it here.

**Review by adammorr — DISMISSED** [review]

File: src/JustSaying/JustSayingBus.cs
Comment (gkinsman) [inline src/JustSaying/JustSayingBus.cs:138]: This no longer protects from multiple-entrants, as multiple calls to this could execute in different threads. Perhaps we need to check for both `_busStarted` and that there are startup tasks? If there are none we could mark as started.

File: tests/JustSaying.IntegrationTests/Fluent/Subscribing/WhenUsingResourceAddresses.cs
Comment (gkinsman) [inline tests/JustSaying.IntegrationTests/Fluent/Subscribing/WhenUsingResourceAddresses.cs:60]: We've been using `Patiently.Until` elsewhere so we don't need explicit waits like this, as it leads to more reliable tests.

File: tests/JustSaying.UnitTests/AwsTools/MessageHandling/Sns/TopicByName/WhenSnsTopicTestBase.cs
Comment (gkinsman) [inline tests/JustSaying.UnitTests/AwsTools/MessageHandling/Sns/TopicByName/WhenSnsTopicTestBase.cs:27]: Is this just to satisfy the rule that IDisposable must be disposed? Wonder if we should just ignore the rule as its a substitute.

File: src/JustSaying/JustSayingBus.cs
Comment (slang25) [reply src/JustSaying/JustSayingBus.cs]: Good spot, I've reverted this and will go again with the suggestion you made

File: tests/JustSaying.UnitTests/AwsTools/MessageHandling/Sns/TopicByName/WhenSnsTopicTestBase.cs
Comment (slang25) [reply tests/JustSaying.UnitTests/AwsTools/MessageHandling/Sns/TopicByName/WhenSnsTopicTestBase.cs:27]: Yeah, I think so, I don't think it hurts but it's a bit noising. I'll push a change with a reduced syntax, there's no need to null these thing out.

**Review by gkinsman — APPROVED** [review]

**Review by adammorr — DISMISSED** [review]

**Review by gkinsman — APPROVED** [review]

**Pull request merged**
